### PR TITLE
Break up stack-arguments, prep v5

### PR DESCRIPTION
**Pull request by pbrisbin**
### [Implement subcommand-specific options](https://github.com/freckle/stack-action/pull/36/commits/66873d2ce0cb83eb63c634d952cc49a8276848c7)
[66873d2](https://github.com/freckle/stack-action/pull/36/commits/66873d2ce0cb83eb63c634d952cc49a8276848c7)

The `stack-arguments` option preceding this was too limited. It could be
used for global arguments to `stack` as well as arguments to `stack
build`, this made it impossible to use the input consistently across all
`stack` commands we happen to invoke.

The new interface separates:

- `stack-arguments` for globals arguments, used in all invocations from
- `stack-build-arguments` for all `stack build` invocations from
- `stack-build-arguments-{step}` for _specific_ `stack build` invocations and, finally
- `stack-{subcommand}-arguments` for other invocations

### [Fix workflows](https://github.com/freckle/stack-action/pull/36/commits/e31a26f5be9669d32c23b01f949e035a24cdc7af)
[e31a26f](https://github.com/freckle/stack-action/pull/36/commits/e31a26f5be9669d32c23b01f949e035a24cdc7af)

- Bump version of actions/cache used
- Bump version of actions/checkout used
- Trim trailing whitespace

### [Use correct name for macOS in comment](https://github.com/freckle/stack-action/pull/36/commits/81b64a1e869d91a2d0556ae92e44cabeaf1902cd)
[81b64a1](https://github.com/freckle/stack-action/pull/36/commits/81b64a1e869d91a2d0556ae92e44cabeaf1902cd)

### [Exclude .git in source-hash](https://github.com/freckle/stack-action/pull/36/commits/c974d191c2e975e2888e19ac3885db8e0db65877)
[c974d19](https://github.com/freckle/stack-action/pull/36/commits/c974d191c2e975e2888e19ac3885db8e0db65877)

We suspect including this is causing some churn in the hash when sources
haven't, in fact, changed.

### [Rewrite README](https://github.com/freckle/stack-action/pull/36/commits/fafd74f0ae081186e4b7f2767b75c23546fc5a00)
[fafd74f](https://github.com/freckle/stack-action/pull/36/commits/fafd74f0ae081186e4b7f2767b75c23546fc5a00)

[Rendered](https://github.com/freckle/stack-action/blob/pb/v5/README.md)

- Describe notable changes in v4 and v5
- Move linter removal notes up to changes subsections
- Tweak generate-matrix README section
- Generate Inputs/Outputs sections

  To do so:

  ```console
  npm instlal
  npm run readme
  ```

- Polish `action.yml`

  - Improve descriptions
  - Don't add `required: true` when there is a default
  - Use `required: false` instead of `default: ""`

  These should all help the generated README look better.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206452824766671

### CODE REVIEW

**Comment by pbrisbin** [comment]
@deemp this PR contains a squashed version of #35 and then corresponding `README` updates. If you don't mind, please review the `README`.

**Comment by deemp** [comment]
I can add a python script that generates Inputs for README from action.yml ([link](https://github.com/nix-community/cache-nix-action/blob/main/scripts/translate_table.py)). Then, I can generate the Inputs section.

Current result:

<table><tr><th>name</th><th>description</th><th>default</th><th>required</th></tr><tr><td><div><p><code>working-directory</code></p></div></td><td><div><p>Working directory for run commands</p></div></td><td><div><p><code>.</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-yaml</code></p></div></td><td><div><p>Override stack.yaml, relative to working-directory</p></div></td><td><div><p><code>stack.yaml</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>test</code></p></div></td><td><div><p>Whether to run tests</p></div></td><td><div><p><code>true</code></p></div></td><td><div><p><code>false</code></p></div></td></tr><tr><td><div><p><code>stack-arguments</code></p></div></td><td><div><p>Additional arguments for all top-level <code>stack</code> command invocations.</p></div></td><td><div><p><code>--no-terminal</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-query-arguments</code></p></div></td><td><div><p>Additional arguments in <code>stack query</code> invocations.</p></div></td><td><div><p><code>''</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-path-arguments</code></p></div></td><td><div><p>Additional arguments in <code>stack path</code> invocations.</p></div></td><td><div><p><code>''</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-setup-arguments</code></p></div></td><td><div><p>Additional arguments in <code>stack setup</code> invocations.</p></div></td><td><div><p><code>''</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-build-arguments</code></p></div></td><td><div><p>Additional arguments for all <code>stack build</code> invocations.</p></div></td><td><div><p><code>--fast --pedantic</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-build-arguments-dependencies</code></p></div></td><td><div><p>Additional arguments passed after <code>stack-build-arguments</code> in <code>stack build</code> invocations on the <code>Dependencies</code> step.</p></div></td><td><div><p><code>''</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-build-arguments-build</code></p></div></td><td><div><p>Additional arguments passed after <code>stack-build-arguments</code> in <code>stack build</code> invocations on the <code>Build</code> step.</p></div></td><td><div><p><code>''</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>stack-build-arguments-test</code></p></div></td><td><div><p>Additional arguments passed after <code>stack-build-arguments</code> in <code>stack build</code> invocations on the <code>Test</code> step.</p></div></td><td><div><p><code>''</code></p></div></td><td><div><p><code>true</code></p></div></td></tr><tr><td><div><p><code>cache-prefix</code></p></div></td><td><div><p>Cache prefix</p></div></td><td><div><p><code>''</code></p></div></td><td><div><p><code>true</code></p></div></td></tr></table>

**Comment by pbrisbin** [comment]
> I can add a python script that generates Inputs for README

Haha, I thought about doing exactly that. Sounds great!

**Comment by deemp** [comment]
scripts/translate_table.py

**Comment by deemp** [comment]
Here's a quick config

scripts/translate_table.py

```py
import yaml

from lxml.etree import tostring, fromstring
from lxml.builder import E
import markdown
from pathlib import Path


def convert_to_table(attrs, is_inputs=True):
    columns = ["name", "description"]

    if is_inputs:
        columns += ["default", "required"]

    root = E.table(
        E.tr(*[E.th(c) for c in columns]),
    )

    for key, value in attrs.items():
        key_pretty = f"`{key}`"
        description = value.get("description", "")
        columns = [key_pretty, description]

        if is_inputs:
            default = (
                f"`{default}`"
                if (default := value.get("default"))
                else "`''`"
                if default == ""
                else ""
            )
            
            required = (
                "`true`"
                if (required := bool(value.get("required", "`false`"))) == True
                else "`false`"
            )

            columns += [
                default,
                required,
            ]

        root.append(
            E.tr(
                *[
                    E.td(fromstring(f"<div>{markdown.markdown(x)}</div>"))
                    for x in columns
                ]
            )
        )

    return (
        tostring(
            root,
            encoding="UTF-16",
            xml_declaration=False,
        )
        .decode("UTF-16")
        .replace("\n", "")
    )


def main():
    actions = [
        ["action.yml", False],
    ]

    for path, has_outputs in actions:
        dir = Path(path).parent
        readme_md = dir / "README.md"
        with open(path, "r") as action:
            action_yaml = yaml.safe_load(action.read())

        with open(readme_md, "r") as readme:
            readme_lines = readme.readlines()

        for i, line in enumerate(readme_lines):
            for attr, heading, is_inputs in [
                ["inputs", "Inputs", True],
                ["outputs", "Outputs", False],
            ]:
                if line.find(f"## {heading}") != -1 and not (
                    heading == "Outputs" and not has_outputs
                ):
                    readme_lines[i + 2] = convert_to_table(
                        action_yaml.get(attr), is_inputs=is_inputs
                    ) + "\n"

        with open(readme_md, "w") as readme:
            readme.writelines(readme_lines)


if __name__ == "__main__":
    main()
```

poetry.toml

```toml
[virtualenvs]
in-project = true
```

pyproject.toml

```toml
[tool.poetry]
name = "scripts"
version = "0.1.0"
description = ""
authors = ["Your Name <you@example.com>"]
license = "MIT"
readme = "README.md"

[tool.poetry.dependencies]
python = "^3.10"
pyyaml = "^6.0.1"
types-pyyaml = "^6.0.12.12"
lxml = "^4.9.3"
markdown = "^3.5.1"

[tool.poetry.scripts]
translate_table = "scripts.translate_table:main"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```

**Comment by pbrisbin** [comment]
Thanks for that! I may use it, or I may just recreate something similar in another scripting language that doesn't require so much... stuff (e.g. the supporting files, etc). Appreciate though, it'd definitely the right idea vs duplicating the details.

**Comment by deemp** [comment]
There's also [yaml-to-table](https://github.com/ishswar/yaml-to-table) that looks more sane

**Comment by pbrisbin** [comment]
I encountered a possible bug intesting this:

```yaml
          stack-build-arguments-test: |
            --test-arguments="--fail-on-focused --print-slow-items=5"
```

When the step runs, the quotes are lost, making it invalid usage,



I will try now with single-quotes, and continue debugging, but FYI.

**Comment by deemp** [comment]
I think it's because of substitution

```sh
echo "stack-build-arguments-test=$(        echo "${{ inputs.stack-build-arguments }} ${{ inputs.stack-build-arguments-test         }}" | tr '\n' ' ')"
```

becomes

```sh
echo "stack-build-arguments-test=$(        echo "--fast --pedantic --test-arguments="--fail-on-focused --print-slow-items=5"" | tr '\n' ' ')"
```

**Comment by deemp** [comment]
I updated the script and made https://github.com/freckle/stack-action/pull/38  just in case.

**Comment by deemp** [comment]
And interpolation should be fixed by https://github.com/freckle/stack-action/pull/39

**Comment by pbrisbin** [comment]
OK, I think this is done and history polished.

**Comment by pbrisbin** [comment]
Here is a PR (private) where I'm testing this branch: https://github.com/freckle/megarepo/pull/32653

**Comment by deemp** [comment]
Would you like to let other people comment on this PR? I can share a link to this PR on a Telegram group (~1.6K Haskellers).

File: package.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "dependencies": {
+    "action-docs": "git+https://github.com/deemp/action-docs.git#0cb0ac2c40736aba24c4fd5097f1074b5a7c7a4b"
```
Comment (deemp) [inline package.json:3]: I believe this package can be moved to `devDependencies`.

File: .github/workflows/example.yml [hunk 2]
```diff
@@ -41,14 +48,17 @@ jobs:
...
         with:
           working-directory: example
-          stack-arguments: --resolver ${{ matrix.resolver }}
-          cache-prefix: ${{ matrix.resolver }}/
-
-      - if: ${{ runner.os != 'windows' }}
+          stack-arguments: --resolver ${{ matrix.stack.resolver }}
+          cache-prefix: ${{ matrix.stack.resolver }}/
+      
+      - if: matrix.stack.ghc
```
Comment (stackptr) [inline .github/workflows/example.yml:54]: Could you add a `name` to describe what this step is doing when run?

File: README.md
Comment (stackptr) [inline README.md:1]: Mind updating line 3 of this README to remove mention of linting?

**Comment by pbrisbin** [comment]
> Would you like to let other people comment on this PR?

It can't hurt. Though I would make it clear the action implementation itself is a mess of shell code, it's just _used_ for Haskell, so it might be more of a review of the interface than the diff, you know?

**Comment by deemp** [comment]
Seems like there are no suggestions from members of that group.

**Comment by pbrisbin** [comment]
Thanks! Sorry this week go away from us, I'll try to merge early last week.

**Review by stackptr — APPROVED** [review]

**Pull request merged**